### PR TITLE
Synchronize `source-mappings.json` from installer too

### PR DESCRIPTION
**Pull request by MilenaHristova**
https://github.com/dotnet/arcade/issues/11982

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/DarcLib/Models/VirtualMonoRepo/SourceMappingFile.cs [hunk 1]
```diff
@@ -19,6 +19,8 @@ public class SourceMappingFile
 
     public string? PatchesPath { get; set; }
 
+    public string? SourceMappings { get; set; }
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/Models/VirtualMonoRepo/SourceMappingFile.cs:22]: Let's do `SourceMappingsPath`

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/InitializeCommandLineOptions.cs [hunk 2]
```diff
@@ -14,6 +14,9 @@ internal class InitializeCommandLineOptions : VmrSyncCommandLineOptions
 {
     [Option('r', "recursive", Required = false, HelpText = $"Process also dependencies (from {VersionFiles.VersionDetailsXml}) recursively.")]
     public bool Recursive { get; set; } = false;
-    
+
+    [Option("source-mappings", Required = true, HelpText = $"A path to the source-mappings.json file to be used for syncing.")]
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/InitializeCommandLineOptions.cs:18]: Either remove `$` or even better, use a constant for "source-mappings.json", I think we have it in `VmrInfo`

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs [hunk 3]
```diff
@@ -71,12 +71,18 @@ public class VmrInitializer : VmrManagerBase, IVmrInitializer
     }
 
     public async Task InitializeRepository(
-        SourceMapping mapping,
+        string mappingName,
         string? targetRevision,
         string? targetVersion,
         bool initializeDependencies,
+        LocalPath sourceMappings,
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrInitializer.cs:78]: I'd include "Path" in this variable name

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 4]
```diff
@@ -141,6 +152,18 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
             return Array.Empty<VmrIngestionPatch>();
         }
 
+        if (_vmrInfo.SourceMappingsPath != null 
+            && _vmrInfo.SourceMappingsPath.StartsWith(VmrInfo.GetRelativeRepoSourcesPath(update.Mapping)))
+        {
+            var fileRelativePath = _vmrInfo.SourceMappingsPath.Remove(0, VmrInfo.GetRelativeRepoSourcesPath(update.Mapping).Length);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:158]: Maybe this?
```suggestion
            var fileRelativePath = _vmrInfo.SourceMappingsPath.Substring(VmrInfo.GetRelativeRepoSourcesPath(update.Mapping).Length);
```

File: src/Microsoft.DotNet.Darc/src/DarcLib/Models/VirtualMonoRepo/SourceMappingFile.cs (hunk 1 above)
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/Models/VirtualMonoRepo/SourceMappingFile.cs:22]: Could I ask you to add XML docs to properties in this class?

You can just copy them from https://github.com/dotnet/dotnet/blob/main/src/source-mappings.json

File: src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/InitializeCommandLineOptions.cs (hunk 2 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/src/Darc/Options/VirtualMonoRepo/InitializeCommandLineOptions.cs:18]: copy-pasting mistake

File: src/Microsoft.DotNet.Darc/src/DarcLib/Models/VirtualMonoRepo/SourceMappingFile.cs (hunk 1 above)
Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/src/DarcLib/Models/VirtualMonoRepo/SourceMappingFile.cs:22]: sure

File: src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/LoadSourceMappingsFromInstallerTest.cs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                Name = Constants.InstallerRepoName,
+                DefaultRemote = InstallerRepoPath,
+                Exclude = new[] { "src/*.dll", "src/*.exe" }
+            }
+        };
+
+        File.WriteAllText(InstallerRepoPath / _sourceMappingsRelativePath,
+            JsonSerializer.Serialize(_sourceMappings, _jsonSettings));
+        File.WriteAllText(InstallerRepoPath / "src" / "excluded.exe", "Excluded exe file");
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/LoadSourceMappingsFromInstallerTest.cs:65]: Please add a comment saying something like "We're excluding new files and adding them into the repo at the same time - they shouldn't be ingested into the VMR"

File: src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs [hunk 6]
```diff
@@ -103,12 +104,30 @@ public class VmrUpdater : VmrManagerBase, IVmrUpdater
...
+            && _vmrInfo.SourceMappingsPath.StartsWith(VmrInfo.GetRelativeRepoSourcesPath(mapping)))
+        {
+            var fileRelativePath = _vmrInfo.SourceMappingsPath.Substring(VmrInfo.GetRelativeRepoSourcesPath(mapping).Length);
+            var clonePath = await _cloneManager.PrepareClone(
+                mapping.DefaultRemote, 
+                targetRevision ?? mapping.DefaultRef, 
+                cancellationToken);
+            
+            _logger.LogDebug($"Loading a new version of source mappings from {clonePath / fileRelativePath}");
+            await _dependencyTracker.InitializeSourceMappings(clonePath / fileRelativePath);
```
Comment (premun) [inline src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:124]: Can we not initialize twice?

Comment (MilenaHristova) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:124]: > Can we not initialize twice?

we need the remote for the clone

Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:124]: We should use SourceManifest for that actually

Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:124]: Hm, on another thought. I think we should do something else.. I will explain this tomorrow

Comment (premun) [reply src/Microsoft.DotNet.Darc/src/DarcLib/VirtualMonoRepo/VmrUpdater.cs:124]: @MilenaHristova let's merge it as-is and I can make additional changes as part of https://github.com/dotnet/arcade/issues/11386

**Review by premun — APPROVED** [review]

**Pull request merged**